src/common/parser.c

```diff
@@ -0,0 +1,399 @@
+		switch (endptr[0]) {
+			case 'K':
+				*size *= 1ULL << 10; /* 1 KB */
+				break;
+			case 'M':
+				*size *= 1ULL << 20; /* 1 MB */
+				break;
+			case 'G':
+				*size *= 1ULL << 30; /* 1 GB */
+				break;
+			case 'T':
+				*size *= 1ULL << 40; /* 1 TB */
+				break;
+		}
+	}
+
+	return PARSER_CONTINUE;
+}
+
+/*
+ * parser_add_part -- (internal) add a new part to the list
+ */
+static void
+parser_add_part(struct headparts *headp, int id, size_t size, char *path)
+{
+	struct part *p = Malloc(sizeof (*p));
+	if (p == NULL) {
+		FATAL("!Malloc");
+	}
+	p->id = id;
+	p->size = size;
+	p->path = path;
+	TAILQ_INSERT_TAIL(headp, p, next);
+}
+
+/*
+ * parser_add_replica -- (internal) add a new replica to the list
+ */
+static struct replica *
+parser_add_replica(struct headreps *headr, int id)
+{
+	struct replica *r = Malloc(sizeof (*r));
+	if (r == NULL) {
+		FATAL("!Malloc");
+	}
+	r->rep_id = id;
+	r->nparts = 0;
+	r->rep_size = 0;
+	TAILQ_INIT(&r->rep_parts);
+	TAILQ_INSERT_TAIL(headr, r, next);
+	return TAILQ_LAST(headr, headreps);
+}
+
+/*
+ * parser_init_poolset -- init a pool set structure
+ */
+void
+parser_init_poolset(struct poolset *ps)
+{
+	TAILQ_INIT(&ps->set_parts);
+	TAILQ_INIT(&ps->reps);
+	ps->nparts = 0;
+	ps->set_size = 0;
+	ps->nreps = 0;
+}
+
+/*
+ * parser_free_poolset -- free a pool set structure
+ */
+void
+parser_free_poolset(struct poolset *ps)
+{
+	struct part *p;
+	struct replica *r;
+
+	/* free pool set parts */
+	while (!TAILQ_EMPTY(&ps->set_parts)) {
+		p = TAILQ_FIRST(&ps->set_parts);
+		TAILQ_REMOVE(&ps->set_parts, p, next);
+		Free(p->path);
+		Free(p);
+	}
+
+	/* free all replicas */
+	while (!TAILQ_EMPTY(&ps->reps)) {
+		r = TAILQ_FIRST(&ps->reps);
+		/* free parts of the current replica */
+		while (!TAILQ_EMPTY(&r->rep_parts)) {
+			p = TAILQ_FIRST(&r->rep_parts);
+			TAILQ_REMOVE(&r->rep_parts, p, next);
+			Free(p->path);
+			Free(p);
+		}
+		TAILQ_REMOVE(&ps->reps, r, next);
+		Free(r);
+	}
+}
+
+/*
+ * parser_parse_set_file -- parse a pool set file
+ */
+int
+parser_parse_set_file(const char *path, int fd, struct poolset *ps)
+{
+	LOG(4, "parsing file %s", path);
+
+#define	REPLICA_HDR_SIG "REPLICA"
+#define	REPLICA_HDR_SIG_LEN 7
+#define	PARSER_MAX_LINE 1024
+
+	enum parser_states_enum {
+		STATE_BEGIN,
+		STATE_PMEMPOOLSET,
+		STATE_REPLICA
+	} parser_state;
+
+	enum parser_codes result;
+	char line[PARSER_MAX_LINE];
+	char *s, *ppath;
+	struct replica *rep = NULL;
+	size_t psize;
+	FILE *fs;
+
+	/* associate a stream with the file descriptor */
+	if ((fs = fdopen(fd, "r")) == NULL) {
+		ERR("!fdopen %s", path);
+		return -1;
+	}
+
+	int nlines = 0;
+	int nparts = 0;
+	size_t set_size = 0;
+	size_t rep_size = 0;
+	parser_state = STATE_BEGIN;
+	result = PARSER_CONTINUE;
+	while (result == PARSER_CONTINUE) {
+		/* read next line */
+		s = fgets(line, PARSER_MAX_LINE, fs);
+		nlines++;
+
+		/* every line starting with '#' is a comment and is skipped */
+		if (s && s[0] == '#') {
+			continue;
```

Comment:
Here you check if `s` is not NULL *AND* if it is a comment line.  So if `s` is NULL you continue on to code that indexes into `line[]`.  Seems like you should check if `fgets()` returns NULL before anything else.

When you do check for comments, how about: `if (*s == '#' || *s == '\n')` so that you skip comments and blank lines.